src/v/kafka/server/group.cc

```diff
@@ -1207,17 +1218,55 @@ void group::insert_prepared(prepared_tx tx) {
 
 ss::future<cluster::commit_group_tx_reply>
 group::commit_tx(cluster::commit_group_tx_request r) {
-    // TODO: https://app.clubhouse.io/vectorized/story/2219
-    // (check for tx_seq to prevent old commit requests committing
-    // ongoing transactions in the same session)
+    // doesn't make sense to fence off a commit because transaction
+    // manager has already decided to commit and acked to a client
+
+    if (_partition->term() != _term) {
+        co_return make_commit_tx_reply(cluster::tx_errc::timeout);
+    }
 
-    auto ongoing_it = _prepared_txs.find(r.pid);
-    if (ongoing_it == _prepared_txs.end()) {
+    auto prepare_it = _prepared_txs.find(r.pid);
+    if (prepare_it == _prepared_txs.end()) {
         vlog(
           klog.trace, "can't find a tx {}, probably already comitted", r.pid);
         co_return make_commit_tx_reply(cluster::tx_errc::none);
     }
 
+    if (prepare_it->second.tx_seq > r.tx_seq) {
+        // rare situation:
+        //   * tm_stm prepares (tx_seq+1)
+        //   * prepare on this group passed but tm_stm failed to write to disk
+        //   * during recovery tm_stm recommits (tx_seq)
+        // existence of {pid, tx_seq+1} implies {pid, tx_seq} is committed
+        vlog(
+          klog.trace,
+          "prepare for pid:{} has higher tx_seq:{} than given: {} => replaying "
+          "already comitted commit",
+          r.pid,
+          prepare_it->second.tx_seq,
+          r.tx_seq);
+        co_return make_commit_tx_reply(cluster::tx_errc::none);
+    } else if (prepare_it->second.tx_seq < r.tx_seq) {
+        if (_recovery_policy == best_effort) {
+            vlog(
+              klog.error,
+              "Rejecting commit with tx_seq:{} since prepare with lesser "
+              "tx_seq:{} exists",
+              r.tx_seq,
+              prepare_it->second.tx_seq);
+            co_return make_commit_tx_reply(cluster::tx_errc::request_rejected);
+        } else {
+            vassert(
+              false,
```

Comment:
nit: you can get rid of the if/else conditional and `vassert(_recovery_policy == best_effort, ..)`

Reply:
even when it's `best_effort` I want to log the error because it's exceptional situation an operator chose to recover from